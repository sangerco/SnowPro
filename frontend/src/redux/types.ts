export interface UserData {
    username: string;
    firstName: string;
    lastName: string;
    email: string;
    avatar: string;
    bio: string;
    videos: string[];
    photos: string[];
    favMountains: string[];
  }


export const FETCH_USER_DATA_REQUEST = 'FETCH_USER_DATA_REQUEST';
export const FETCH_USER_DATA_SUCCESS = 'FETCH_USER_DATA_SUCCESS';
export const FETCH_USER_DATA_FAILURE = 'FETCH_USER_DATA_FAILURE';


export interface FetchUserDataRequestAction {
    type: typeof FETCH_USER_DATA_REQUEST
}

export interface FetchUserDataSuccessAction {
    type: typeof FETCH_USER_DATA_SUCCESS;
    payload: UserData;
}

export interface FetchUserDataFailureAction {
    type: typeof FETCH_USER_DATA_FAILURE;
    payload: string;
}

export type UserActionTypes = 
    FetchUserDataRequestAction
    | FetchUserDataSuccessAction
    | FetchUserDataFailureAction;