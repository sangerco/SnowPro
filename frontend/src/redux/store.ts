import { Action, compose } from "redux";
import { ThunkAction, configureStore } from "@reduxjs/toolkit";
import { logger } from "redux-logger";
import rootReducer from "./reducers/rootReducer";

const composeEnhancers =
  (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(logger, composeEnhancers),
});
