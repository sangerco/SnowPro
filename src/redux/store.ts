import { configureStore, combineReducers } from "@reduxjs/toolkit";
import userReducer from "./slices/userSlice";
import authReducer from "./slices/authSlice";
import messageReducer from "./slices/messageSlice";
import messageReplyReducer from "./slices/messageReplySlice";
import reviewReducer from "./slices/reviewSlice";
import reviewReplyReducer from "./slices/reviewReplySlice";
import tagReducer from "./slices/tagSlice";
import storage from "redux-persist/lib/storage";
import { persistReducer, persistStore } from "redux-persist";

const persistConfig = {
  key: "root",
  storage,
  blacklist: ["messages", "messageReplies", "reviews", "reviewReplies"],
};

const rootReducer = combineReducers({
  users: userReducer,
  auth: authReducer,
  messages: messageReducer,
  messageReplies: messageReplyReducer,
  reviews: reviewReducer,
  reviewReplies: reviewReplyReducer,
  tags: tagReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const persistor = persistStore(store);
