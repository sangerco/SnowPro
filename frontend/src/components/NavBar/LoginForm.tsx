import React, { useState } from 'react';
import { connect } from 'react-redux';
import { sendLoginData } from '../../redux/actions/userActions';
import { RootState } from '../../redux/store';
import { Button, Input, Form, Message } from 'semantic-ui-react'

interface LoginData {
    username: string,
    password: string
};

interface LoginProps {
    user: LoginData | null;
    loading: boolean;
    error: string;
    sendLoginData: (
      username: string, 
      password: string) => Promise<void>
};

const LoginForm: React.FC<LoginProps> = ({ user, loading, error, sendLoginData }) => {
    const [ formData, setFormData ] = useState<LoginData>({
        username: '',
        password: ''
    });

    const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();

        await sendLoginData(formData.username, formData.password);

        setFormData({ username: '', password: ''});
    };

    const handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
        const { name, value } = e.target;
        setFormData({
            ...formData,
            [name]: value
        });
    };

    return (
        <Form onSubmit={handleSubmit} error>
            <Form.Field>
                <label>username</label>
                <input placeholder='username' name='username' value={formData.username} onChange={handleChange}/>
            </Form.Field>
            <Form.Field>
                <label>Choose Password</label>
                <Input type='password' name='password' value={formData.password} onChange={handleChange}/>
            </Form.Field>
            <Button type='submit'>Login</Button>
            {loading && <p>Loading...</p>}
            <Message
                error
                header={error}
                content={error}
            />
        </Form>
    );
}

const mapStateToProps = (state: RootState) => ({
    newUser: state.user.data,
    loading: state.user.loading,
    error: state.user.error
})
  
  export default connect(mapStateToProps, { sendLoginData })(LoginForm);