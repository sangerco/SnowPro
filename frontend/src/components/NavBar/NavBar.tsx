import React from 'react';
import { connect, useDispatch, useSelector } from 'react-redux';
import { logoutUser } from '../../redux/actions/userActions';
import { Link } from 'react-router-dom';
import { Menu, Dropdown } from 'semantic-ui-react';
import { RootState } from '../../redux/store';
import { useAuth } from '../AuthProvider';

interface NavBarProps {
    logoutUser: () => void;
}

const NavBar: React.FC<NavBarProps> = ({ logoutUser }) => {
    const dispatch = useDispatch();
    const { username } = useAuth();

    const isAuthenticated = useSelector((state: RootState) => state.login.token)

    const handleLogout = () => {
        dispatch(logoutUser() as any);
    };

    const userMenuOptions = [
        { key: 'profile', text: 'Profile Page', icon: 'user', as: Link, to: `/users/${username}` },
        { key: 'inbox', text: 'Messages', icon: 'envelope', as: Link, to: `/users/${username}/inbox`},
        { key: 'userlist', text: 'Users', icon: 'users', as: Link, to: `/users` },
        { key: 'logout', text: 'Logout', icon: 'sign-out', onClick: handleLogout},
    ];

    return (
        <Menu inverted color='black' size='large'>
            <Menu.Item header as={Link} to='/'>
                App Logo TBD
            </Menu.Item>

            <Menu.Item as={Link} to='/ski-areas'>
                All Ski Areas
            </Menu.Item>

            {isAuthenticated ? (
                <Menu.Menu position='right'>
                    <Dropdown item text='User Menu' options={userMenuOptions} />
                </Menu.Menu>
            ) : (
                <Menu.Menu position='right'>
                    <Menu.Item as={Link} to='/login'>
                        Login
                    </Menu.Item>
                    <Menu.Item as={Link} to='/register'>
                        Register
                    </Menu.Item>
                </Menu.Menu>
            )}
        </Menu>
    )
};

const mapStatetoProps = (state: RootState) => ({
    user: state.logout.user
});

const mapDispatchToProps = {
    logoutUser
};

export default connect(mapStatetoProps, mapDispatchToProps)(NavBar);