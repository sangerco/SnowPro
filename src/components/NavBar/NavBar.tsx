import React, { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { RootState, AppDispatch } from "../../redux/store";
import { logoutUser } from "../../redux/slices/authSlice";
import { Menu, Dropdown, Button, Image } from "semantic-ui-react";
import { Link, useNavigate } from "react-router-dom";
import logo from "../../media/snow-pro-logo.jpg";

const NavBar = () => {
  const auth = useSelector((state: RootState) => state.auth);
  const dispatch = useDispatch<AppDispatch>();
  const navigate = useNavigate();

  const isAuthenticated = auth.isAuthenticated;

  const handleLogout = () => {
    dispatch(logoutUser());

    navigate("/");
  };

  return (
    <Menu inverted color="black" size="large">
      <Menu.Item header as={Link} to="/">
        <Image src={logo} alt={"snow-pro-logo"} size="small" />
      </Menu.Item>
      <Dropdown item text="Check it out" style={{ color: "goldenrod" }}>
        <Dropdown.Menu>
          <Dropdown.Item>
            <Link to="/ski-areas" style={{ color: "green" }}>
              Ski Areas
            </Link>
          </Dropdown.Item>
          <Dropdown.Item>
            <Link
              to="/ski-areas/reviews/recent-reviews"
              style={{ color: "green" }}>
              Recent Reviews
            </Link>
          </Dropdown.Item>
          {isAuthenticated ? (
            <Dropdown.Item>
              <Link to="/users/all-users" style={{ color: "green" }}>
                Users
              </Link>
            </Dropdown.Item>
          ) : null}
        </Dropdown.Menu>
      </Dropdown>
      {isAuthenticated ? (
        <Menu.Menu position="right">
          <Menu.Item>
            <Link to="/" style={{ color: "goldenrod" }}>
              Home
            </Link>
          </Menu.Item>
          <Menu.Item onClick={handleLogout} style={{ color: "goldenrod" }}>
            Logout
          </Menu.Item>
        </Menu.Menu>
      ) : (
        <Menu.Menu position="right">
          <Menu.Item>
            <Link to="/login" style={{ color: "goldenrod" }}>
              Login
            </Link>
          </Menu.Item>
          <Menu.Item>
            <Link to="/register" style={{ color: "goldenrod" }}>
              Sign Up!
            </Link>
          </Menu.Item>
        </Menu.Menu>
      )}
    </Menu>
  );
};

export default NavBar;
