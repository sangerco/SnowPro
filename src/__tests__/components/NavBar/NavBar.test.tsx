/* eslint-disable testing-library/no-node-access */
import React from "react";
import NavBar from "../../../components/NavBar/NavBar";
import { render, fireEvent, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../../redux/store";
import { BrowserRouter } from "react-router-dom";

describe("test Nav Bar", () => {
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <NavBar />
        </BrowserRouter>
      </Provider>
    );
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <NavBar />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
