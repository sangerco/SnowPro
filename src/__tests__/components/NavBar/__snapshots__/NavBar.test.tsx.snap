// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Nav Bar should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui black large inverted menu"
  >
    <a
      class="header item"
      href="/"
    >
      <img
        alt="snow-pro-logo"
        class="ui small image"
        src=""
      />
    </a>
    <div
      aria-expanded="false"
      class="ui item dropdown"
      role="listbox"
      style="color: goldenrod;"
      tabindex="0"
    >
      <div
        aria-atomic="true"
        aria-live="polite"
        class="divider text"
        role="alert"
      >
        Check it out
      </div>
      <i
        aria-hidden="true"
        class="dropdown icon"
      />
      <div
        class="menu transition"
      >
        <div
          class="item"
          role="option"
        >
          <a
            href="/ski-areas"
            style="color: green;"
          >
            Ski Areas
          </a>
        </div>
        <div
          class="item"
          role="option"
        >
          <a
            href="/ski-areas/reviews/recent-reviews"
            style="color: green;"
          >
            Recent Reviews
          </a>
        </div>
      </div>
    </div>
    <div
      class="right menu"
    >
      <div
        class="item"
      >
        <a
          href="/login"
          style="color: goldenrod;"
        >
          Login
        </a>
      </div>
      <div
        class="item"
      >
        <a
          href="/register"
          style="color: goldenrod;"
        >
          Sign Up!
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
